### PR TITLE
feat: verify built tx against intended tx

### PR DESCRIPTION
**Pull request by xstelea**

### CODE REVIEW

File: src/application/transaction/verifyTx.ts [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+  (txIntent: TransactionIntent, trackError: TrackError) =>
+  (tx: BuiltTransaction) => {
+    return Transaction.fromBuffer(Buffer.from(tx.transaction.blob, 'hex'))
+      .map(tx => {
+        log.info(tx.toString())
+        return tx.instructions
+      })
+      .map(filterUpAndEndInstructions)
+      .map(groupInstructionsByActions)
+      .map(parseGroupData(Network.STOKENET))
```
Comment (AbstractFruitFactory) [inline src/application/transaction/verifyTx.ts:285]: hardcoded network?

File: src/application/transaction/verifyTx.ts [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+          }
+
+          default:
+            return ok(null)
+        }
+      }),
+    ),
+  )
+
+const assert = (condition: boolean, errorMessage: string) =>
```
Comment (AbstractFruitFactory) [inline src/application/transaction/verifyTx.ts:145]: nice!

File: src/application/transaction/verifyTx.ts [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    `incorrect amount expected '${expected.toString()}' got '${actual.toString()}'`,
+  )
+
+const assertResource = (expected: ResourceIdentifierT, actual: string) => {
+  const expectedHash = expected.hash.toString('hex')
+  return assert(
+    expectedHash === actual,
+    `incorrect rri expected '${expectedHash}' got '${actual}'`,
+  )
+}
```
Comment (AbstractFruitFactory) [inline src/application/transaction/verifyTx.ts:189]: nice!

File: src/application/transaction/verifyTx.ts [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    assertNumberOfInstructions(2, instructions.length),
+    assertSubstateTypes(
+      [SubStateType.TOKENS],
+      instructions.map(({ substateType }) => substateType),
+    ),
+    assertAccountAddress('from', action.from_account, from.owner),
+    assertAccountAddress('to', action.to_account, to.owner),
+    assertResource(action.rri, from.resource),
+    assertResource(action.rri, to.resource),
+    isSelfTransfer ? ok(null) : assertAmount(action.amount, to.amount),
```
Comment (AbstractFruitFactory) [inline src/application/transaction/verifyTx.ts:208]: this, for some reason, messes up the return type of the `Result`.

fix: 

```suggestion
    isSelfTransfer ? ok<null, string>(null) : assertAmount(action.amount, to.amount),
```

File: src/application/transaction/verifyTx.ts [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    assertNumberOfInstructions(2, instructions.length),
+    assertSubstateTypes(
+      [SubStateType.TOKENS, SubStateType.PREPARED_STAKE],
+      instructions.map(({ substateType }) => substateType),
+    ),
+    assertAccountAddress('from', action.from_account, from.owner),
+    assertAccountAddress('to', action.to_validator, to.validator),
+    assertResource(action.rri, from.resource),
+    assertAmount(action.amount, to.amount),
+  ])
```
Comment (AbstractFruitFactory) [inline src/application/transaction/verifyTx.ts:227]: On these functions, you can flatten the `null` type like:

```suggestion
  ]).map(flattenNullArray)
  
  const flattenNullArray = (n: null[]) => n[0]
```

File: src/application/transaction/verifyTx.ts [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+        groups.filter((group): group is InstructionGroup => group.length > 0),
+      )
+      .andThen(verifyTxIntent(txIntent))
+      .map(() => tx)
+      .mapErr(errors => {
+        trackError({
+          errors: [
+            radixAPIError({
+              message: 'transaction intent does not match built transaction',
+              details: { type: 'VerifyBuiltTransactionError', errors },
```
Comment (AbstractFruitFactory) [inline src/application/transaction/verifyTx.ts:295]: The `errors` variable here can have several types. Mostly problematic might be the fact that we can expect `string[]` but also `Error`.

We probably want to change the `string` error types to `Error` to be consistent with the return type of `Transaction.fromBuffer`, and then properly display the messages using something like `errors.map(e => e.message)`.

Comment (xstelea) [reply src/application/transaction/verifyTx.ts]: Good idea. Changed string to Error

**Review by AbstractFruitFactory — APPROVED** [review]

**Pull request merged**